### DISCUSSION TITLE
ORM for notion databases

### DISCUSSION THREAD

**Discussion by aahnik** [ramnes/notion-sdk-py#36] [Ideas]
Django ORM, sqlalchemy etc, are examples where you write python classes and call python functions, but under the hood SQL queries get carried out.

It would be great to have an ORM, that under the hood would carry out various interactions using Notion API. For the end user, it will be a high level experience.

Example usage

```python

# a vague example usage

class Student(NotionBaseModel):
    name: Title
    address: Text
    last_edited: LastEditedTime

student = Student("Aahnik","India") # will add a row to the table
```

**Comment by Jeadie** [comment]
I think you will find this is very similar to the implementation I am working on using dataclasses for easy instantiation and access.

**Comment by ramnes** [comment]
I'm quite biased because I already wrote a kind of ORM and had a lot of fun with it (namely [Mongo-Thingy](https://github.com/numberly/mongo-thingy)), but yeah, I would love this. :)

Let's land @Jeadie's work first (#10) but feel free to give ideas on the overall API here in the meantime!

**Comment by jheddings** [comment]
I made a very simple draft implementation of an ORM for Notion SDK.  It is modeled loosely after SQLAlchemy and `notion-py`.  It is very much still a work in progress.  I also extended the result iterators to automatically return custom typed objects if requested.

You can see the work in progress here:
https://github.com/jheddings/notion-sdk-py/tree/records/notion_client

This allows a user to create custom objects:

```
class Book(Page):
    Title = Property('Title')
    Subtitle = Property('Subtitle')
    Publisher = Property('Publisher')
    PageCount = Property('Page Count', type=types.Number)
    ReleaseDate = Property('Release Date', type=types.Date)
    AgeRating = Property('Age Rating', type=types.SelectOne)
```

Using this approach, we could easily add the write-back-to-Notion methods, I just haven't gotten that far yet.

**Reply by jheddings** [reply]
I've taken this about as far I am planning to prove out the idea until Jeadie's PR goes through.  If anyone wants to partner on this to improve the design or work more details, I'm happy to open up the branch.

This branch is working to declare objects and update values on the server.  See the example for details.  Note that I've also merged the content iterators into this branch, assuming that PR goes through eventually.

https://github.com/jheddings/notion-sdk-py/tree/records

**Comment by jheddings** [comment]
Just FYI - I've created a new "high level" module for Notion built on the core SDK.  Check out [notional](https://github.com/jheddings/notional) if you are interested.  I'd welcome collaborators, especially people smarter than I am for this 😊
